### PR TITLE
Spec: Add Privacy and Security considerations sections

### PR DESCRIPTION
**Pull request by alexmturner**
Adds two non-normative sections discussing privacy and security considerations

***
<a href="https://pr-preview.s3.amazonaws.com/patcg-individual-drafts/private-aggregation-api/pull/53.html" title="Last updated on Jun 9, 2023, 7:58 PM UTC (906cd62)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/patcg-individual-drafts/private-aggregation-api/53/3d32307...906cd62.html" title="Last updated on Jun 9, 2023, 7:58 PM UTC (906cd62)">Diff</a>

### CODE REVIEW

**Comment by alexmturner** [comment]
@linnan-github, could you PTAL? Thanks!
cc @yoavweiss

File: spec.bs [hunk 1]
```diff
@@ -676,3 +676,190 @@ To <dfn>determine the numeric value</dfn> of a [=signal base value=]
...
+
+This API lets isolated contexts with access to cross-site data (i.e. <a
+href="https://wicg.github.io/shared-storage/">Shared Storage</a> worklets/<a
+href="https://wicg.github.io/turtledove/">Protected Audience</a> script runners)
+send aggregatable reports over the network.
+
+Aggregatable reports contain encrypted high entropy cross-site information, in
+the form of key-value pairs (i.e. contributions to a histogram). The information
+embedded in the contributions is arbitrary but can include things like browsing
+history and other cross-site activity. The API aims to provide some protection
```
Comment (yoavweiss) [inline spec.bs:696]: "some protection" sounds weaker than what I think this API actually provides

File: spec.bs [hunk 2]
```diff
@@ -676,3 +676,190 @@ To <dfn>determine the numeric value</dfn> of a [=signal base value=]
...
+for this information from being passed from one site to another.
+
+### Restricted contribution processing ### {#restricted-contribution-processing}
+
+The histogram contributions are not exposed directly. Instead, they are
+encrypted so that they can only be processed by a trusted aggregation service.
+This trusted aggregation service sums the values across the reports for each key
+and adds noise to each of these values to produce ‘summary reports’.
+
+The output of that processing will be an aggregate, noised histogram. The
```
Comment (yoavweiss) [inline spec.bs:706]: s/aggregate/aggregated/ ?

File: spec.bs [hunk 3]
```diff
@@ -676,3 +676,190 @@ To <dfn>determine the numeric value</dfn> of a [=signal base value=]
...
+parameters are left [=implementation-defined=] to allow different and evolving
+choices in the tradeoffs between information exposure and utility.
+
+Clearing site data {#clearing-site-data}
+----------------------------------------
+
+The [=aggregatable report cache=] as well as any contribution history data
+stored for the [=PrivateAggregation/consume budget if permitted=] algorithm
+contain data about a user’s web activity. As such, the user agent is strongly
+encouraged to expose controls that allow the user to delete data from them.
```
Comment (yoavweiss) [inline spec.bs:745]: I'd prefer a MUST language in the normative parts of the spec. (similar to https://wicg.github.io/client-hints-infrastructure/#accept-ch-cache-definition)

File: spec.bs (hunk 1 above)
Comment (alexmturner) [reply spec.bs]: Rephrased this to "aims to protect"

File: spec.bs (hunk 3 above)
Comment (alexmturner) [reply spec.bs]: The privacy considerations section was marked as non-normative, but I've switched this part to be normative (and use "must"). Not sure if it makes sense to switch the key rotation recommendation (under "Protecting the histogram contributions") to a normative "should" as well.

Comment (yoavweiss) [reply spec.bs]: I think it'd be better to move these MUSTs and SHOULDs to the normative parts that define the cache, and then have the P&S section non-normative and just clarify that these concerns are handled by those normative sections.

Comment (alexmturner) [reply spec.bs]: Makes sense -- moved the normative clearing data section under 'Storage'

File: spec.bs [hunk 4]
```diff
@@ -676,3 +683,190 @@ To <dfn>determine the numeric value</dfn> of a [=signal base value=]
...
+### Restricted contribution processing ### {#restricted-contribution-processing}
+
+The histogram contributions are not exposed directly. Instead, they are
+encrypted so that they can only be processed by a trusted aggregation service.
+This trusted aggregation service sums the values across the reports for each key
+and adds noise to each of these values to produce ‘summary reports’.
+
+The output of that processing will be an aggregated, noised histogram. The
+service ensures that any report can not be processed multiple times. Further,
+information exposure is limited by contribution budgets on the client. In
```
Comment (linnan-github) [inline spec.bs:715]: Shall we use `on the user agent` here?

File: spec.bs [hunk 5]
```diff
@@ -676,3 +683,190 @@ To <dfn>determine the numeric value</dfn> of a [=signal base value=]
...
+cross-site data. The recipient of the report may also be able to observe
+side-channel information such as the time when the report was sent, or IP
+address of the sender.
+
+### Protecting against leaks via the number of reports ### {#protecting-against-leaks-via-the-number-of-reports}
+
+However, the number of reports with the given metadata could expose some
+cross-site information. To protect against this, the API delays sending reports
+by a randomized amount of time to make it difficult to determine whether a
+report was sent or not from any particular event. In the case that a context ID
```
Comment (linnan-github) [inline spec.bs:731]: Could we add an issue to spec the null report generation with context ID?

File: spec.bs [hunk 6]
```diff
@@ -676,3 +683,190 @@ To <dfn>determine the numeric value</dfn> of a [=signal base value=]
...
+    effectively hides the reporting origin from the network administrator, but
+    is likely impractical to enforce and is itself perhaps circumventable by the
+    network administrator, e.g. by monitoring IP addresses instead.
+
+### User-presence tracking ### {#user-presence-tracking}
+
+The browser only tries to send reports while it is running and while it has
+internet connectivity (even without an explicit check for connectivity,
+naturally the report will fail to be sent if there is none), so receiving or not
+receiving an aggregatable report at the expected time leaks information about
```
Comment (linnan-github) [inline spec.bs:810]:
```suggestion
receiving an [=aggregatable report=] at the expected time leaks information about
```

Comment (linnan-github) [inline spec.bs:810]: The aggregatable report is delayed by a random time, so the "expected time" may be a bit unclear?

File: spec.bs [hunk 7]
```diff
@@ -676,3 +683,190 @@ To <dfn>determine the numeric value</dfn> of a [=signal base value=]
...
+reporting origin, including at-home vs. at-work or approximate real-world
+geolocation, or reveal patterns in the user’s browsing activity.
+
+Possible mitigations include:
+
+1. Send reports immediately: This effectively eliminates the presence tracking,
+    as the original request made to the reporting origin is in close temporal
+    proximity to the report request. However, it increases the likelihood that
+    the reporting origin can correlate the original API invocation to the report
+    being sent, which weakens the privacy controls of the API, see [Protecting
```
Comment (linnan-github) [inline spec.bs:822]: The link seems broken?

File: spec.bs [hunk 8]
```diff
@@ -303,6 +303,13 @@ Storage {#storage}
 A user agent holds an <dfn>aggregatable report cache</dfn>, which is a [=list=]
 of [=aggregatable reports=].
 
+Clearing storage {#clearing-storage}
+----------------------------------------
+
+The user agent must expose controls that allow the user to delete data from
```
Comment (linnan-github) [inline spec.bs:309]: Do we want to mention why the contribution cache data is not cleared? Or should they?

File: spec.bs (hunk 5 above)
Comment (alexmturner) [reply spec.bs:730]: Updated an issue to be clearer

File: spec.bs (hunk 6 above)
Comment (alexmturner) [reply spec.bs]: Linked, but put "(serialized)" too as the spec's struct doesn't exactly correspond to what the server receives

File: spec.bs (hunk 8 above)
Comment (alexmturner) [reply spec.bs:308]: That's a good point -- I think we don't usually require this data to be deleted as it's short-lived, but I'm not sure if there's explicit guidance on this. Added a comment along these lines and said the user agent "may" provide controls for them. I'm not sure what's typical here though.

File: spec.bs [hunk 9]
```diff
@@ -676,3 +685,213 @@ To <dfn>determine the numeric value</dfn> of a [=signal base value=]
...
+The length of the payload could additionally expose some cross-site information,
+namely how many contributions are included. To protect against this, the payload
+will be padded in the future.
+
+Issue(56): Pad the payload to avoid this risk.
+
+### Temporary debugging mechanism ### {#temporary-debugging-mechanism}
+
+The `enableDebugMode()` method allows for the protections of this API to be
+bypassed to ease testing and integration. Specifically, the contents of the
```
Comment (linnan-github) [inline spec.bs:750]: Are all the protections bypassed? Or shall we say weakened?

Comment (alexmturner) [reply spec.bs]: Moved to "allows for many of the protections"

**Review by linnan-github — APPROVED** [review]

**Comment by alexmturner** [comment]
Thanks!

**Pull request merged**